Add documentation about Wagtail page models (#4707)
cfpb/cfgov-refresh: docs/wagtail-pages.md

# Wagtail pages

[Wagtail pages](http://docs.wagtail.io/en/v1.13.4/topics/pages.html) are 
[Django models](https://docs.djangoproject.com/en/1.11/topics/db/models/) 
that are constructed of 
[fields](#fields), [StreamFields](#streamfields), and [panels](#panels) 
that are rendered in a standard way. 
All CFPB Wagtail pages should inherit from the 

> [`v1.models.base.CFGOVPage` class](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py#L60).

Reviewer: I'm not sure of a better way to do this, but the `#L60` is likely to get out of date here. Potentially you could link to a particular commit, but I'm not sure that's better.

Author: Yeah, I'm not sure. Maybe if we generated API docs for CFGOVPage it'd be better (ala #4706).

Reviewer: Another possible thought would be to clean up [models/base.py](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py) so that the file itself is a satisfactory thing to link to, without the page number. Currently most of that file is CFGOVPage, with a few other odds and ends that could be relocated.

Author: For now I'll just link to `models/base.py`... that gets anyone who's curious close enough I think.

> There are types of information defined on a new Wagtail page model: 

Reviewer: Missing word "three" here?

Reviewer 2: 'twasn't.

```suggestion
There are three types of information defined on a new Wagtail page model: 
```

basic [database fields](#fields) (like any Django model), 
specialized database fields called [StreamFields](#streamfields) that allow for freeform page content, 
and [editor panels](#panels) that present these fields to content editors. 

## Fields

Database fields in Wagtail pages work exactly the same as in 
[Django models](https://docs.djangoproject.com/en/1.11/topics/db/models/#fields), 
and Wagtail pages can use any [Django model field](https://docs.djangoproject.com/en/1.11/ref/models/fields/). 

For example, our 

> [`BrowsePage` includes a standard Django `BooleanField`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/browse_page.py#L54) 

Reviewer: See above re: `#L54`.

They allow a content editor to pick any number number of optional components
and place them in any order within their StreamField. 
In practice this provides the 
[flexibilty of a large rich text box with the structure of individual components](https://torchbox.com/blog/rich-text-fields-and-faster-horses/).

For example, our [`LandingPage` page model includes a `header` StreamField](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/landing_page.py#L13) that can have a hero and/or a text introduction:

```python
from wagtail.wagtailcore.fields import StreamField

from v1.atomic_elements import molecules
from v1.models import CFGOVPage

class LandingPage(CFGOVPage):

    header = StreamField([
        ('hero', molecules.Hero()),
        ('text_introduction', molecules.TextIntroduction()),
    ], blank=True)
```

**Insert link to StreamField docs here.**

## Panels

Editor panels define how the page's [fields](#fields) and [StreamFields](#streamfields) will be organized for content editors.

The base Wagtail `Page` class and the [`v1.models.base.CFGOVPage` class](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py#L60) 

> defines a specific set of panels to which all fields should be added:

Reviewer: Is the term "panels" one that the average developer or editor would understand? I tend to think of these as "tabs" on the page edit screen.

Author: Yeah, I describe where each panel appears below, but maybe stating it here too would be good. I also thought about including screenshots, but that seemed a bit superfluous.

- `content_panels`. 
  These fields appear on the "General Content" tab when editing a page. 
  For page body content.

> - [`sidefoot_panels`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py#L114-L116).

Reviewer: Minor: use '`sidefoot_panels`:' instead of '`sidefoot_panels`.'?

  These fields appear on the "Sidebar" tab when editing a page. 
- [`settings_panels`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py#L118-L125).
  Appears on the "Configuration" tab when editing a page. 
  Page configuration such as the categories, tags, scheduled publishing, etc.

Most fields will simply require a `FieldPanel` to be added to one of the sets of panels above. 
StreamFields will require a `StreamFieldPanel`. 
See the [Wagtail documentation for additional, more complex panel options](https://docs.wagtail.io/en/v1.13.4/topics/pages.html#editor-panels).

> For example, in our `BrowsePage` (used in the [database fields example above](#fields),

Reviewer: Missing closing parenthesis.

[the `secondary_nav_exclude_sibling_pages` `BooleanField` is added to the `sidefoot_panels` as a `FieldPanel`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/browse_page.py#L63):

```python
from django.db import models
from wagtail.wagtailadmin.edit_handlers import FieldPanel

from v1.models.base import CFGOVPage

class BrowsePage(CFGOVPage):
    secondary_nav_exclude_sibling_pages = models.BooleanField(default=False)

    # …

    sidefoot_panels = CFGOVPage.sidefoot_panels + [
        FieldPanel('secondary_nav_exclude_sibling_pages'),
    ]

> ```

Reviewer: It would be helpful to have a sentence here about what this actually does, e.g. "This creates a checkbox title 'Secondary nav exclude...' on the "Sidebar" tab in the Wagtail admin. Checking or unchecking this box sets the value of the `secondary_nav_exclude_sibling_pages` field when the page model is saved to the database."

> In out `LandingPage` (used in the [StreamFields example above](#streamfields), 

Reviewer: Typo: "out".

[the `header` StreamField is added to the `content_panels` as a `StreamFieldPanel`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/landing_page.py#L31):

```python
from wagtail.wagtailadmin.edit_handlers import StreamFieldPanel
from wagtail.wagtailcore.fields import StreamField

from v1.atomic_elements import molecules
from v1.models import CFGOVPage

class LandingPage(CFGOVPage):
    header = StreamField([
        ('hero', molecules.Hero()),
        ('text_introduction', molecules.TextIntroduction()),
    ], blank=True)

    # …

    content_panels = CFGOVPage.content_panels + [
        StreamFieldPanel('header'),
        # …
    ]
```

## Parent / child page relationships

> Wagtail provides two attributes to page models that enable restricting the types of subpages or parent pages a particular page model can have. 

Reviewer: Link to [Wagtail docs](https://docs.wagtail.io/en/v1.13.4/topics/pages.html#parent-page-subpage-type-rules) here?

For example, in our [interactive regulations page models](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/models/pages.py#L138) 
we have a `RegulationLandingPage` that can be created anywhere in the page tree. 
`RegulationLandingPage`, however, can only have two types of pages created within it: 
`RegulationPage` and `RegulationSearchPage`. 
This parent/child relationship is expressed by setting `subpage_types` on `RegulationLandingPage`
and `parent_page_types` on `RegulationPage` and `RegulationSearchPage` 
to either the model classes themselves or to a model name in the form `app_label.ModelName`:

```python
from v1.models import CFGOVPage

class RegulationLandingPage(CFGOVPage):
    subpage_types = ['regulations3k.RegulationPage', 'RegulationsSearchPage']

class RegulationsSearchPage(CFGOVPage):
    parent_page_types = ['regulations3k.RegulationLandingPage']
    subpage_types = []

class RegulationPage(CFGOVPage):
    parent_page_types = ['regulations3k.RegulationLandingPage']
    subpage_types = []
```

!!! note 
    Note: by setting `subpage_types` to an empty list on `RegulationPage` and `RegulationSearchPage`, 

>     we ensure that neither of these pages can have child pages.

Reviewer: This wording is a little confusing. Perhaps instead "we prevent child pages from being added to these page types."

> ## Template rendering

Reviewer: Link to [Wagtail docs](https://docs.wagtail.io/en/v1.13.4/topics/pages.html#template-rendering) somewhere here?

Author: Linked in 2b0f4a2.

Sometimes new Wagtail page types will need to make customizations to their base template when rendering the page.
This is done by overriding the `template` attribute on the page model.

For example, the [interactive regulations landing page](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/models/pages.py#L138) 
includes a customized list of recently issued notices that gets loaded dynamically from the Federal Register. 
To do this it provides its own template that inherits from our base templates 
and overrides the `content_sidebar` block to include a seperate `recent_notices` template:

```python
from v1.models import CFGOVPage

class RegulationLandingPage(CFGOVPage):
    template = 'regulations3k/landing-page.html'
```

And [`regulations3k/landing-page.html`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/jinja2/regulations3k/landing-page.html#L5):

```jinja2
{% extends 'layout-2-1-bleedbar.html' %}

{% import 'recent-notices.html' as recent_notices with context %}

{% block content_sidebar scoped -%}
    {{ recent_notices }}
{%- endblock %}

> {% block javascript scoped %}

Reviewer: This block is a little overwhelming and doesn't add much to the example. How would you feel about excluding it? The `content_sidebar` override demonstrates what is being described above.

Author: Excluded in 2b0f4a2.

# Wagtail pages

[Wagtail pages](http://docs.wagtail.io/en/v1.13.4/topics/pages.html) are 
[Django models](https://docs.djangoproject.com/en/1.11/topics/db/models/) 
that are constructed of 
[fields](#fields), [StreamFields](#streamfields), and [panels](#panels) 
that are rendered in a standard way. 
All CFPB Wagtail pages should inherit from the 
[`v1.models.base.CFGOVPage` class](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py).

!!! note
    Before creating a new Wagtail page type 
    please consider whether one of our existing page types can meet your needs, 
    whether this is a different type of content than we have anywhere else in the site, 
    and whether there is maintenance efficiency gained from a new page type.

There are types of information defined on a new Wagtail page model: 
basic [database fields](#fields) (like any Django model), 
specialized database fields called [StreamFields](#streamfields) that allow for freeform page content, 

> and [editor panels](#panels) that present these fields to content editors. 

Reviewer 2: I think it'd be nice to make this a numbered list, but I don't feel strongly about it.

Author: I prefer it as a sentence, personally.

## Fields

Database fields in Wagtail pages work exactly the same as in 
[Django models](https://docs.djangoproject.com/en/1.11/topics/db/models/#fields), 
and Wagtail pages can use any [Django model field](https://docs.djangoproject.com/en/1.11/ref/models/fields/). 

For example, our `BrowsePage` 
[includes a standard Django `BooleanField`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/browse_page.py) 
that allows content editors to toggle secondary navigation sibling pages:

```python
from django.db import models

> from v1.models.base import CFGOVPage

Reviewer 2: It seems a bit unnecessary to include the imports here when you're already heavily subsetting the class definition, but I defer to your experience and expertise.

Author: I like to have the relevant imports. `models` is shown and used, and `CFGOVPage` is too.

class BrowsePage(CFGOVPage):
    secondary_nav_exclude_sibling_pages = models.BooleanField(default=False)
```

## StreamFields

StreamFields are special Django model fields provided by Wagtail for 
[freeform page content](https://docs.wagtail.io/en/v1.13.4/topics/streamfield.html). 
They allow a content editor to pick any number number of optional components
and place them in any order within their StreamField. 

> In practice this provides the flexibilty of a large 

Reviewer 2: ```suggestion
In practice, this provides the flexibility of
```

> [rich text box with the structure of individual components](https://torchbox.com/blog/rich-text-fields-and-faster-horses/).

Reviewer 2: ```suggestion
[a large rich text field, with the structure of individual components](https://torchbox.com/blog/rich-text-fields-and-faster-horses/).
```

For example, our `LandingPage` page model 
[includes a `header` StreamField](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/landing_page.py) 
that can have a hero and/or a text introduction:

```python
from wagtail.wagtailcore.fields import StreamField

from v1.atomic_elements import molecules
from v1.models import CFGOVPage

class LandingPage(CFGOVPage):

    header = StreamField([
        ('hero', molecules.Hero()),
        ('text_introduction', molecules.TextIntroduction()),
    ], blank=True)
```

The specifics of StreamField block components can be found in 
[Creating and Editing Wagtail Components](https://cfpb.github.io/cfgov-refresh/editing-components/).

## Panels

> Editor panels define how the page's [fields](#fields) and [StreamFields](#streamfields) will be organized for content editors 

Reviewer 2: ```suggestion
Editor panels define how the page's [fields](#fields) and [StreamFields](#streamfields) will be organized for content editors;
```

> and correspond to the tabs that appear across the top of the edit view for pages in the Wagtail admin.

Reviewer 2: ```suggestion
they correspond to the tabs that appear across the top of the edit view for a page in the Wagtail admin.
```

Reviewer 2: This sentence kinda muddies the waters a bit... the tabs are not the panels; the tabs contain one or more panels. But it may not be important to make that distinction here.

> The base Wagtail `Page` class and the [`v1.models.base.CFGOVPage` class](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py#L60) 

Reviewer 2: Seems unnecessary to repeat the full module path to `CFGOVPage` here, and might be useful to be explicit that it's a subclass of `Page`.

```suggestion
The base Wagtail `Page` class and the [`CFGOVPage` subclass of it](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/base.py) 
```

> define a specific set of panels to which all fields should be added:

Reviewer 2: ```suggestion
define three specific sets of panels to which all fields should be added:
```

Author: It's actually not three, it's four. `Page` has `promote_panels`, but we subsume `promote_panels` into `settings_panels`.

Reviewer 2: OK, well, either way, it's multiple panel sets, not one.

Author: Changed to "define specific sets" in afe72bc.

- `content_panels`:
  For page body content.
  These fields appear on the "General Content" tab when editing a page. 
- `sidefoot_panels`:
  For page sidebar or footer content.
  These fields appear on the "Sidebar" tab when editing a page. 
- `settings_panels`:
  Page configuration such as the categories, tags, scheduled publishing, etc.
  Appears on the "Configuration" tab when editing a page. 

Most fields will simply require a `FieldPanel` to be added to one of the sets of panels above. 
StreamFields will require a `StreamFieldPanel`. 
See the [Wagtail documentation for additional, more complex panel options](https://docs.wagtail.io/en/v1.13.4/topics/pages.html#editor-panels).

For example, in our `BrowsePage` (used in the [database fields example above](#fields)),
the `secondary_nav_exclude_sibling_pages` `BooleanField` 

> [is added to the `sidefoot_panels` as a `FieldPanel`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/browse_page.py#L63):

Reviewer 2: Continuing the removal of likely-to-get-outdated line references.

```suggestion
[is added to the `sidefoot_panels` as a `FieldPanel`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/browse_page.py):
```

```python
from django.db import models
from wagtail.wagtailadmin.edit_handlers import FieldPanel

from v1.models.base import CFGOVPage

class BrowsePage(CFGOVPage):
    secondary_nav_exclude_sibling_pages = models.BooleanField(default=False)

    # …

    sidefoot_panels = CFGOVPage.sidefoot_panels + [
        FieldPanel('secondary_nav_exclude_sibling_pages'),
    ]
```

> Because `secondary_nav_exclude_sibling_page` is a boolean field, 

Reviewer 2: ```suggestion
Because `secondary_nav_exclude_sibling_pages` is a boolean field, 
```

> this creates a checkbox on the "Sidebar" tab when editing a page.

Reviewer 2: ```suggestion
this creates a checkbox on the "Sidebar/Footer" tab when editing a page.
```

> Checking or unchecking that checkbox will set the value of `secondary_nav_exclude_sibling_pages` when the page is saved.

Reviewer 2: ```suggestion
Checking or unchecking that checkbox will set the value of `secondary_nav_exclude_sibling_pages` when the page is saved.
```

Author: What's the change here? I don't see it in the suggestion.

Reviewer 2: I guess this was a mistake.

> In our `LandingPage` (used in the [StreamFields example above](#streamfields), 

Reviewer 2: ```suggestion
In our `LandingPage` (used in the [StreamFields example above](#streamfields)), 
```

the `header` StreamField 
[is added to the `content_panels` as a `StreamFieldPanel`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/landing_page.py#L31):

```python
from wagtail.wagtailadmin.edit_handlers import StreamFieldPanel
from wagtail.wagtailcore.fields import StreamField

from v1.atomic_elements import molecules
from v1.models import CFGOVPage

class LandingPage(CFGOVPage):
    header = StreamField([
        ('hero', molecules.Hero()),
        ('text_introduction', molecules.TextIntroduction()),
    ], blank=True)

    # …

    content_panels = CFGOVPage.content_panels + [
        StreamFieldPanel('header'),
        # …
    ]
```

## Parent / child page relationships

Wagtail provides two attributes to page models that enable 

> [restricting the three types of subpages or parent pages](https://docs.wagtail.io/en/v1.13.4/topics/pages.html#parent-page-subpage-type-rules) 

Reviewer 2: Errant "three" here?

Author: FIxed in afe72bc.

a particular page model can have. 

> For example, in our [interactive regulations page models](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/models/pages.py#L138) 

Reviewer 2: ```suggestion
For example, in our [interactive regulations page models](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/models/pages.py)
```

we have a `RegulationLandingPage` that can be created anywhere in the page tree. 
`RegulationLandingPage`, however, can only have two types of pages created within it: 
`RegulationPage` and `RegulationSearchPage`. 
This parent/child relationship is expressed by setting `subpage_types` on `RegulationLandingPage`
and `parent_page_types` on `RegulationPage` and `RegulationSearchPage` 

> to either the model classes themselves or to a model name in the form `app_label.ModelName`:

Reviewer 2: This took me a minute to understand. Can we pick a preferred style and just say that?

Author: What do you mean? setting both `subpage_types` and `parent_page_types` is required.

Author: I'll add the two bullets from the Wagtail docs.

Reviewer 2: I'm referring to the specific line I commented on: "to either the model classes themselves or to a model name in the form `app_label.ModelName`:". Can we pick one to prefer?

Reviewer 2: Silly Django. Is there a reason that both methods need to exist, yet neither be preferred?

And would it be good to include the single quotes around `'app_label.ModelName'` for clarity?

Author: Yes, sometimes you can't import the module class directly, or sometimes you need to reference it before it's been defined. That's what the string-based lookup is for.

We could just recommend the string. I don't have strong feelings one way or the other, but both are possible and valid in the Django universe.

Author: Done in 15edfd8.

```python
from v1.models import CFGOVPage

class RegulationLandingPage(CFGOVPage):
    subpage_types = ['regulations3k.RegulationPage', 'RegulationsSearchPage']

class RegulationsSearchPage(CFGOVPage):
    parent_page_types = ['regulations3k.RegulationLandingPage']
    subpage_types = []

class RegulationPage(CFGOVPage):
    parent_page_types = ['regulations3k.RegulationLandingPage']
    subpage_types = []
```

!!! note 

>     Note: we prevent child pages from being added to `RegulationPage` and `RegulationSearchPage` 

Reviewer 2: "Note" is already in the header of the note.

```suggestion
    We prevent child pages from being added to `RegulationPage` and `RegulationSearchPage` 
```

    by setting `subpage_types` to an empty list.

## Template rendering

> Sometimes new Wagtail page types will need to make customizations to their base template 

Reviewer 2: Does this ever _not_ happen?

```suggestion
New Wagtail page types will usually need to make customizations to their base template 
```

Author: With a subclass of `Page`, you're right. With `CFGOVPage`, I'm honestly not sure. I like your wording better though.

Reviewer 2: Well, `CFGOVPage` doesn't have its own template definition, so anything subclassing it has to provide its own, right? I guess if you're subclassing a _subclass_ of CFGOVPage, you may not need to. But if you're doing that and not needing to change the template, do you _really_ need to do that?

[when rendering the page](https://docs.wagtail.io/en/v1.13.4/topics/pages.html#template-rendering).
This is done by overriding the `template` attribute on the page model.

> For example, the [interactive regulations landing page](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/models/pages.py#L138) 

Reviewer 2: ```suggestion
For example, the [interactive regulations landing page](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/models/pages.py) 
```

includes a customized list of recently issued notices that gets loaded dynamically from the Federal Register. 
To do this it provides its own template that inherits from our base templates 
and overrides the `content_sidebar` block to include a seperate `recent_notices` template:

```python
from v1.models import CFGOVPage

class RegulationLandingPage(CFGOVPage):
    template = 'regulations3k/landing-page.html'
```

> And [`regulations3k/landing-page.html`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/jinja2/regulations3k/landing-page.html#L5):

Reviewer 2: ```suggestion
And in [`regulations3k/landing-page.html`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/jinja2/regulations3k/landing-page.html):
```

> ```jinja2

Reviewer 2: Unfortunately, `jinja2` is not a supported syntax.

The GH suggestion feature doesn't work in this situation, but I suggest just removing it and not replacing it with anything else.

Author: Eh, what if it is supported at some future date? Just because it's not useful in code rendering doesn't mean it's not necessarily useless in the doc (I read Markdown for fun sometimes :P ).

Reviewer 2: Well, it's _counterproductive_ to the rendering if there is actual HTML in the snippet, but since this case is exclusively Jinja tags, I suppose that's fine.

        ('text_introduction', molecules.TextIntroduction()),
    ], blank=True)

    # …

    content_panels = CFGOVPage.content_panels + [
        StreamFieldPanel('header'),
        # …
    ]
```

## Parent / child page relationships

Wagtail provides two attributes to page models that enable 
[restricting the types of subpages or parent pages](https://docs.wagtail.io/en/v1.13.4/topics/pages.html#parent-page-subpage-type-rules) 
a particular page model can have. On any page model:

- `parent_page_types` limits which page types this type can be created under.
- `subpage_types` limits which page types can be created under this type.

For example, in our [interactive regulations page models](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/regulations3k/models/pages.py#L138) 
we have a `RegulationLandingPage` that can be created anywhere in the page tree. 
`RegulationLandingPage`, however, can only have two types of pages created within it: 
`RegulationPage` and `RegulationSearchPage`. 
This parent/child relationship is expressed by setting `subpage_types` on `RegulationLandingPage`
and `parent_page_types` on `RegulationPage` and `RegulationSearchPage` 

> to a model name in the form `app_label.ModelName`:

Reviewer 2: Same question about including the single quotes.

Reviewer 2: (Same question as in the conversation about recommending the string method. Unrelated to the first question in this review.)

Author: Yeah, it should be a regular string. I'll remove the backticks.

Author: Done in 9ceb770.

> to a model name in the form 'app_label.ModelName':

Reviewer 2: I mean, I think you want the quoted string in backticks, still 😆

Author: But that just looks weird. If it's a string, shouldn't it render with the rest of the text, instead of being encapsulated as a string encapsulated in code?

Reviewer 2: I guess that depends on your view of how code formatting backticks should be used. I would do it, personally, but it's not a big deal.

```python
from v1.models import CFGOVPage

class RegulationLandingPage(CFGOVPage):

>     subpage_types = ['regulations3k.RegulationPage', 'RegulationsSearchPage']

Reviewer 2: Should this say `'regulations3k.RegulationsSearchPage'`?

Author: You know, it probably should. It doesn't in the source code for the example, but I might change the example because Good Practice.

Author: Done in 9ceb770.
